src/populate.ts

```diff
@@ -0,0 +1,357 @@
+              ...typeFragments[t.name].map(({ fragment }) => {
+                // Add used fragment for insertion at Document node
+                getUsedFragments(fragment).forEach(f =>
+                  requiredUserFragments.add(userFragments[f])
+                );
+
+                // Add fragment for insertion at Document node
+                additionalFragments = {
+                  ...additionalFragments,
+                  [fragment.name.value]: fragment,
+                };
+
+                return {
+                  kind: 'FragmentSpread',
+                  name: {
+                    kind: 'Name',
+                    value: fragment.name.value,
+                  },
+                } as const;
+              }),
+            ];
+          }, [] as FragmentSpreadNode[]);
+
+          const existingSelections =
+            (node.selectionSet && node.selectionSet.selections) || [];
+
+          const selections =
+            existingSelections.length + newSelections.length !== 0
+              ? [...newSelections, ...existingSelections]
+              : [
+                  {
+                    kind: 'Field',
+                    name: {
+                      kind: 'Name',
+                      value: '__typename',
+                    },
+                  },
+                ];
+
+          return {
+            ...node,
+            directives,
+            selectionSet: {
+              kind: 'SelectionSet',
+              selections,
+            },
+          };
+        },
+      },
+      Document: {
+        leave: node => {
+          return {
+            ...node,
+            definitions: [
+              ...node.definitions,
+              ...Object.keys(additionalFragments).map(
+                k => additionalFragments[k]
+              ), // Object.values
+              ...Array.from(requiredUserFragments),
+            ],
+          };
+        },
+      },
+    })
+  );
+};
+
+/** Get all possible types for node with TypeInfo. */
+const getTypes = (schema: GraphQLSchema, typeInfo: TypeInfo) => {
+  const typeInfoType = typeInfo.getType();
+  const type =
+    isListType(typeInfoType) || isNonNullType(typeInfoType)
+      ? typeInfoType.ofType
+      : typeInfoType;
+
+  if (!isCompositeType(type)) {
+    warn(
+      `PopulateExchange: Unsupported type "${type}" at populate decorator.`,
+      1
+    );
+    return [];
+  }
+
+  if (isInterfaceType(type) || isUnionType(type)) {
+    return schema.getPossibleTypes(type);
+  }
+
+  return [type];
+};
+
+/** Get name of non-abstract type for adding to 'typeFragments'. */
+const getTypeName = (t: TypeInfo) => {
+  const typeInfoType = t.getType();
+  const type =
+    isListType(typeInfoType) || isNonNullType(typeInfoType)
+      ? typeInfoType.ofType
+      : typeInfoType;
+
+  if (isAbstractType(type)) {
+    invariant(
+      'PopulateExchange',
+      'Invalid TypeInfo state. "getTypeName" does not expect to receive an abstract type.',
+      1
```

Comment:
Still an error code that isn't in `help.md` but I'll fix that up in the branch I've mentioned, which I'll have to rebase anyway :)